src/dotty/tools/dotc/ast/untpd.scala

```diff
@@ -94,17 +94,51 @@ object untpd extends Trees.Instance[Untyped] with UntypedTreeInfo {
   class InfixOpBlock(leftOperand: Tree, rightOp: Tree) extends Block(leftOperand :: Nil, rightOp)
 
   // ----- Modifiers -----------------------------------------------------
+  /** Mod is intended to record syntactic information about modifiers, it's
+    * NOT a replacement of flagsets.
+    *
+    * For any query about semantic information, check `flags` instead.
+    */
+  sealed trait Mod extends Positioned
 
- /** Modifiers and annotations for definitions
-   *  @param flags          The set flags
+  object Mod {
+    case class Private() extends Mod
```

Comment:
@liufengyun I think this can be accommodated in addFlag:

      if (compatible(mods.flags, flag)) mods | mod
      else mods.withAddedMod(mod)